updatehub-sdk/src/listener.rs

```diff
@@ -0,0 +1,140 @@
+// Copyright (C) 2018, 2019, 2020 O.S. Systems Sofware LTDA
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::Result;
+use async_ctrlc::CtrlC;
+use async_std::{
+    io::BufReader,
+    os::unix::net::{UnixListener, UnixStream},
+    prelude::*,
+};
+use std::{collections::HashMap, net::Shutdown, process};
+
+const SDK_TRIGGER_FILENAME: &str =
+    "/usr/share/updatehub/state-change-callbacks.d/10-updatehub-sdk-statechange-trigger";
+const SOCKET_PATH: &str = "updatehub-statechange.sock";
+
+type StateChangeCallback = fn(State);
+type ErrorCallback = fn(String);
+
+pub struct StateChangeListener {
+    listeners: HashMap<String, StateChangeCallback>,
+    error_handlers: Vec<ErrorCallback>,
+}
+
+pub enum StateID {
+    Download,
+    Install,
+    Reboot,
+    Error,
+}
+
+pub struct State {
+    id: StateID,
+    conn: UnixStream,
+}
+
+impl State {
+    pub fn downloading(&mut self) {
+        self.conn.write_all(b"downloading");
+    }
+
+    pub fn rebooting(&mut self) {
+        self.conn.write_all(b"rebooting");
+    }
```

Comment:
These messages should never be *written to the socket*. The SDK read states names and can either respond with `cancel` or not respond to proceed (both implemented correctly below).